cranelift-codegen/src/legalizer/split.rs

```diff
@@ -134,11 +134,7 @@ pub fn split_ebb_params(func: &mut ir::Function, cfg: &ControlFlowGraph, ebb: Eb
     let mut repairs = Vec::new();
     let pos = &mut FuncCursor::new(func).at_top(ebb);
 
-    for (num, ebb_param) in pos
-        .func
-        .dfg
-        .ebb_params(ebb)
-        .to_vec()
+    for (num, ebb_param) in SmallVec::<[Value; 16]>::from_slice(pos.func.dfg.ebb_params(ebb))
```

Comment:
I've tried to see if we could just not materialize the vector here, by iterating over the ebb_params, but this causes ownerships issues with "pos" which aren't trivial to fix. So we need to materialize it unfortunately.

I've looked at this code a bit more for context, and this should only serve to break i128 into smaller types, as the "continue" statement a few lines below suggests. It's unfortunate that we create this vector even if there's no i128 types in the graph, as should be the case in our wasm test suite. A couple alternatives are available here:

## Filtering out

- take an iter out of the ebb_params
- filter_map the ebb_param only when their type is i128
- then collect into a SmallVec

Then, in the general wasm case, the filtered iterator will be empty, and *hopefully* rustc is good enough to not even allocate the SmallVec in this case.

## Skipping if there's no i128 type among ebb params

So adding a check at the top of this function by iterating over ebb_params and if !any of them has the i128 type, early return. This has better memory usage certainty to it, but is less convenient to maintain (e.g. if we ever add support for other types -- would i64 need to be handled here for x86 i686? we'll see...).

Reply:
I redid it using the "skipping" scheme.  It works out pretty nicely, avoiding all allocation in the non-splitting case, and with a reduced instruction count too, in my test case.